src/itwinai/components.py

```diff
@@ -88,47 +88,58 @@
 from abc import ABC, abstractmethod
 import time
 import functools
+import builtins as __builtin__
 # import logging
 # from logging import Logger as PythonLogger
 
 from .type import MLModel, MLDataset, MLArtifact
 from .serialization import ModelLoader, Serializable
 from .distributed import (
     detect_distributed_environment,
-    distributed_patch_print
+    distributed_patch_print,
+    builtin_print
 )
 
 
-def monitor_exec(method: Callable) -> Callable:
+def monitor_exec(print_on_main_worker_only: bool = False) -> Callable:
     """Decorator for execute method of a component class.
     Computes execution time and gives some information about
     the execution of the component.
 
     Args:
-        func (Callable): class method.
+        print_on_main_worker_only (bool): whether to ignore ``print()`` calls on
+            workers which are not the main worker in a distributed context,
+            such as distributed ML. If set to ``True``, to force printing on
+            all workers you will need to use ``print(..., force=True)``.
     """
-    @functools.wraps(method)
-    def monitored_method(self: BaseComponent, *args, **kwargs) -> Any:
-        # Disable print in workers different from the main one,
-        # when in distributed environments.
-        dist_grank = detect_distributed_environment().global_rank
-        distributed_patch_print(is_main=dist_grank == 0)
-
-        msg = f"Starting execution of '{self.name}'..."
-        self._printout(msg)
-        start_t = time.time()
-        try:
-            # print(f'ARGS: {args}')
-            # print(f'KWARGS: {kwargs}')
-            result = method(self, *args, **kwargs)
-        finally:
-            self.cleanup()
-        self.exec_t = time.time() - start_t
-        msg = f"'{self.name}' executed in {self.exec_t:.3f}s"
-        self._printout(msg)
-        return result
-
-    return monitored_method
+
+    def method_decorator(method: Callable) -> Callable:
+        @functools.wraps(method)
+        def wrapper(self: BaseComponent, *args, **kwargs) -> Any:
+            if print_on_main_worker_only:
+                # Disable print in workers different from the main one,
+                # when in distributed environments.
+                dist_grank = detect_distributed_environment().global_rank
+                distributed_patch_print(is_main=dist_grank == 0)
+
+            msg = f"Starting execution of '{self.name}'..."
+            self._printout(msg)
+            start_t = time.time()
+            try:
+                result = method(self, *args, **kwargs)
+            finally:
+                self.cleanup()
+            self.exec_t = time.time() - start_t
+            msg = f"'{self.name}' executed in {self.exec_t:.3f}s"
+            self._printout(msg)
+
+            # Reset print() to __builtins__ one
+            __builtin__.print = builtin_print
```

Comment:
I am kind of new to this syntax here, but wouldn't this either 
1) Reset what you did in the if statement or
2) Repeat what you did in the if statement (thus giving no value)?

Not sure what applies in terms of scope etc. here :)

Reply:
it will reset it, avoiding to tamper with prints in subsequent code. We only want to ignore prints on workers for the component's logic

Reply:
I don't think I am entirely following. Wouldn't that negate the stuff you did on line 119 and below?

Reply:
yes

Reply:
Could you not just use something similar to "strategy.is_main_worker" inside of this function to do this, instead of overwriting the built-in print function? Since it is only for this scope anyway, why overwrite the global function?

Reply:
what you suggest would be possible only when strategy is available, which is not always true for all components unfortunately. You may also mean "why not checking if strategy.is_main_worker every time before printing": this is possible but tedious and error prone. The new version of the code should solve both problems providing two decorators, one to monitor the execution of a component, the other to suppress the prints. It is now much more straightforward

Reply:
I see now that I misunderstood what you were trying to do. I didn't consider the fact that you are calling the component before resetting, so I thought you were suppressing only the 2-3 prints where you are saying that training started. Now I see that you are also suppressing all the calls in the component as well, and so now I fully agree with what you are doing.